capstone/src/main/java/com/google/sps/data/RecommenderAlgorithm.java

```diff
@@ -11,49 +13,158 @@
-        PreparedQuery otherUserResults = datastore.prepare(otherUserQuery);
- 
-        User otherUser = null;
-        HashMap<String, Integer> otherUserTagMap = null;
-        HashMap<User, HashMap<String,Integer>> otherUserVectorMap = new HashMap<User, HashMap<String,Integer>>();
-        for(Entity otherUserEntity : otherUserResults.asIterable()) {
-            otherUser = new User(otherUserEntity);
-            otherUserTagMap = RecommenderAlgorithm.populateUserTagMap(otherUser);
-            otherUserVectorMap.put(otherUser, otherUserTagMap);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+
+        if (results.countEntities() == 0) {
+            return new ArrayList<>();
         }
- 
-        // Algorithm is work in progress. Have yet to normalize and calculate relationship.
-        return null;
+
+        for (Entity result : results.asIterable()) {
+            Room room = new Room(result);
+            if (room.getAllFollowers().contains(key) && !room.getAllFollowers().contains(user.getUserKey()) &&
+                !rooms.contains(room.getRoomKey())) {
+                rooms.add(room.getRoomKey());
+            }
+        }
+        return rooms;
     }
- 
-    public static HashMap<String, Integer> populateUserTagMap(User user) {
+
+    public static void populateEmbeddedEntity(User user) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        Filter userRoomFilter = new FilterPredicate(Room.FOLLOWERS_PROPERTY_KEY, FilterOperator.EQUAL, user.getUserKey());
-        Query userRoomQuery = new Query(Room.ROOM_ENTITY_NAME).setFilter(userRoomFilter);
-        PreparedQuery userRoomResults = datastore.prepare(userRoomQuery);
- 
-        HashMap<String, Integer> userTagMap = Tag.constructUserVectorMap();
-        List<String> userTags = null;
-        for(Entity roomEntity : userRoomResults.asIterable()){
-            userTags = new Room(roomEntity).getAllTags();
-            for(String tag : userTags) {
-                userTagMap.put(tag, userTagMap.get(tag) + 1);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+        EmbeddedEntity entity = user.getEmbeddedTags();
+
+        if (results.countEntities() == 0) {
+            return;
+        }
+
+        entity = Tag.constructUserEmbeddedEntity();
+        for (Entity result : results.asIterable()) {
+            Room room = new Room(result);
+            if (room.getAllFollowers().contains(user.getUserKey())) {
+                for (String tag : room.getAllTags()) {
+                    double frequency = (Double) entity.getProperty(tag);
+                    entity.setProperty(tag, frequency + 1.0);
+                }
             }
         }
+        user.setEmbeddedTags(entity);
+    }
+
+    public static void normalizeEmbeddedEntity(User user) {
```

Comment:
Should this be public? As the interface stands, I this is only used by this class

Reply:
no it should be private you are right